File: go/0202E-快乐数.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-16 15:41:24
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func isHappy(n int) bool {
	vis := make(map[int]bool)

	getSum := func(x int) int {
		sum := 0
		for x > 0 {
			sum += (x % 10) * (x % 10)
			x /= 10
		}
		return sum
	}

	for n != 1 {
		if vis[n] {
			return false
		}
		vis[n] = true
		n = getSum(n)
	}
	return true
}

File: cpp/0202E-快乐数.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-14 08:57:09
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    bool isHappy(int n) {
        unordered_set<int> vis;
        while (1) {
            int s = sum(n);
            if (s == 1) return true;
            if (vis.find(s) != vis.end())
                return false;
            else
                vis.insert(s);
            n = s;
        }
    }

    int sum(int n) {
        int res = 0, t;
        while (n) {
            t = n % 10;
            res += t * t;
            n /= 10;
        }
        return res;
    }
};